src/Formats/FormatSchemaFactory.cpp

```diff
@@ -0,0 +1,220 @@
+#include <Formats/FormatFactory.h>
+#include <Formats/FormatSchemaFactory.h>
+#include <Formats/FormatSchemaInfo.h>
+#include <google/protobuf/compiler/parser.h>
+#include <google/protobuf/descriptor.pb.h>
+#include <google/protobuf/io/tokenizer.h>
+#include <google/protobuf/io/zero_copy_stream_impl.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int AMBIGUOUS_FORMAT_SCHEMA;
+extern const int FORMAT_SCHEMA_ALREADY_EXISTS;
+extern const int UNKNOWN_FORMAT_SCHEMA;
+extern const int UNKNOWN_FORMAT_SCHEMA_TYPE;
+extern const int INVALID_DATA;
+}
+
+namespace
+{
+String basename(const std::filesystem::path & path)
+{
+    return path.filename().replace_extension().string();
+}
+
+String getSchemaTypeFromFileExtension(String ext) noexcept
+{
+    /// when we support more schema format types, we probably should do it
+    /// in a better way and consolidate it with FormatSchemaInfo.
+    if (ext == ".proto")
+        return "Protobuf";
+    else
+        return "";
+}
+
+class ErrorCollector final: public google::protobuf::io::ErrorCollector
+{
+private:
+    std::vector<String> errors;
+
+public:
+    void AddError(int line, google::protobuf::io::ColumnNumber column, const std::string & message) override
+    {
+        std::ostringstream error;
+        error << "line: " << line << ", column: " << column << ", error: " << message;
+        errors.push_back(error.str());
+    }
+
+    String toString() const
+    {
+        std::ostringstream ret;
+        for (const auto & error : errors)
+            ret << error << std::endl;
+        return ret.str();
+    }
+};
+
+void validateProtobufSchema(const String & payload)
+{
+    std::istringstream s_payload(payload);
+    google::protobuf::io::IstreamInputStream input(&s_payload);
+    ErrorCollector error_collector;
+    google::protobuf::io::Tokenizer tokenizer(&input, &error_collector);
+    google::protobuf::FileDescriptorProto descriptor;
+    google::protobuf::compiler::Parser parser;
+
+    parser.RecordErrorsTo(&error_collector);
+    if (!parser.Parse(&tokenizer, &descriptor))
+        throw Exception(ErrorCodes::INVALID_DATA, "Invalid schema, errors: \n{}", error_collector.toString());
+}
+}
+
+void FormatSchemaFactory::registerSchema(const ContextPtr & context, const String & schema_name, const String & schema_type, const String & schema_body, ExistsOP if_exists)
+{
+    assert(!schema_name.empty());
+    assert(!schema_type.empty());
+    assert(!schema_body.empty());
+
+    checkSchemaType(schema_type);
+
+    auto format_settings = getFormatSettings(context);
+    FormatSchemaInfo fsinfo{schema_name, schema_type, false, true, format_settings.schema.format_schema_path};
+
+    std::lock_guard lock(mutex);
```

Comment:
We should make the lock granularity as small as possible

Reply:
I think this is the smallest scope, suggestions?